lib/uploader.rb

```diff
@@ -3,21 +3,34 @@
 require 'pry'
 require 'json'
 require_relative './chipmunk_bag'
-
-CHIPMUNK_URL = "http://localhost:3000"
+require_relative './chipmunk_service'
+require_relative './bag_rsyncer'
 
 class Uploader
-  def initialize(api_key,bag_path)
-    @api_key = api_key
+  def initialize(api_key,bag_path,service: ChipmunkService.new(api_key: api_key),rsyncer: BagRsyncer.new(bag_path))
     @bag_path = bag_path.chomp('/')
     @request_params = request_params_from_bag(bag_path)
+    @service = service
+    @rsyncer = rsyncer
   end
 
   def upload
-    req = make_request
-    rsync_bag(req["upload_link"])
-    qitem = complete_request(req)
-    print_result(wait_for_bag(qitem))
+    begin
+      req = make_request
+
+      if !req["upload_link"]
```

Comment:
Here, uploader checks the result of the request for the upload link, and does something if it fails. To me, this seems like the only place where we should detect if the bag has already been uploaded (via checking if there is a storage location), because it is the use case. I suggest removing the check from Bag, and to just do it here. That change should make it more clear within uploader why we're refusing to upload a bag. Right now, it knows a lot about Bag's implementation.